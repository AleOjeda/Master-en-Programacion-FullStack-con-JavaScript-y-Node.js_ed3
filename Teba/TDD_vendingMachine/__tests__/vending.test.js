const { maquinaExpendedora } = require('../vending')
const { seedClients } = require('../fixtures/clients')
const { seedProducts } = require('../fixtures/products')
const config = require('../fixtures/config')

let productos, clientes = [];
let vendingMachine = {};


beforeEach(() => {
    productos = seedProducts();
    clientes = seedClients();
    vendingMachine = maquinaExpendedora(config, clientes, productos);
})

describe('methods for managing clients and their needs', () => {
    // Get all clients
    test('must bring all the clients', () => {
        expect(vendingMachine.getAllClients()).toBe(clientes)
    })
    // Test isClient
    test('must confirm existing clients', () => {
        expect(vendingMachine.isClient("Inventado")).toBeFalsy()
        expect(vendingMachine.isClient(clientes[0].usuario)).toBeTruthy()
    })
    // Add client
    test('must create new client', () => {
        expect(vendingMachine.getAllClients().length).toBe(2)
        vendingMachine.addClient({
            nombre: "Teba Gomez",
            usuario: "teba",
            contraseña: "pass2",
            tipo: "usuario",
            presupuesto: "100",
            gasto: []})
        expect(vendingMachine.getAllClients().length).toBe(3)
        expect(vendingMachine.getAllClients()).toMatchSnapshot()
      })
    // Avoid adding existing clients
    test('must avoid adding existing clients',() =>{
        expect(vendingMachine.getAllClients().length).toBe(2)
        vendingMachine.addClient(clientes[0])
        expect(vendingMachine.getAllClients().length).toBe(2)
        expect(vendingMachine.getAllClients()).toMatchSnapshot()
    })
    // Avoid adding clients with incomplete info
    test('must avoid adding clients with incomplete info', () => {
        expect(() => vendingMachine.addClient({})).toThrow("Missing Client Info!")
        expect(vendingMachine.getAllClients().length).toBe(2)
        expect(vendingMachine.getAllClients()).toMatchSnapshot()
    })
    // Remove client
    test('must remove specific client', () => {
        expect(vendingMachine.getAllClients().length).toBe(2)
        expect(vendingMachine.removeClient("ulises")).toBeTruthy()
        expect(vendingMachine.getAllClients().length).toBe(1)
        expect(vendingMachine.getAllClients()).toMatchSnapshot()
    })
    // Avoid remove non-existing clients
    test('must avoid removing non-existnig clients', () => {
        expect(vendingMachine.getAllClients().length).toBe(2)
        expect(vendingMachine.removeClient("Inventado")).toBeFalsy()
        expect(vendingMachine.getAllClients().length).toBe(2)
        expect(vendingMachine.getAllClients()).toMatchSnapshot()
    })
    
})


/*
### Paso 3 - Creamos varios métodos para gestionar a los clientes y sus necesidades
Metodos:
- Agregar un cliente √
- Evitaremos que se registren usuarios que ya existan √
- Evitaremos dar de alta usuarios que no esten debidamente cumplimentados √
- Eliminar un cliente √
- Evitaremos eliminar usuarios que no existan √√√√√√√√

- Consultar saldo de un cliente
    - Comprobaremos la contraseña y el usuario
    - Devolverá el saldo existente o -1 en caso de error
- Consultar gastos de un cliente
    - Comprobaremos la contraseña y el usuario
    - Devolverá un array o false en caso de error

Caracteristicas:
- Protegeremos las funciones de gestión de usuarios con la contraseña ficticiaMola
*/