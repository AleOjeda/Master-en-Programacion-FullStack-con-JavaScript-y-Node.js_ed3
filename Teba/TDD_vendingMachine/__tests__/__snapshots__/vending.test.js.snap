// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods for managing clients and their needs must avoid adding clients with incomplete info 1`] = `
Array [
  Object {
    "contraseña": "pass",
    "gasto": Array [],
    "nombre": "Ulises Gascón",
    "presupuesto": "1000",
    "tipo": "admin",
    "usuario": "ulises",
  },
  Object {
    "contraseña": "pass_2",
    "gasto": Array [],
    "nombre": "Carlos Perez",
    "presupuesto": "100",
    "tipo": "usuario",
    "usuario": "cPerez",
  },
]
`;
exports[`methods for managing clients and their needs must avoid adding existing clients 1`] = `
Array [
  Object {
    "contraseña": "pass",
    "gasto": Array [],
    "nombre": "Ulises Gascón",
    "presupuesto": "1000",
    "tipo": "admin",
    "usuario": "ulises",
  },
  Object {
    "contraseña": "pass_2",
    "gasto": Array [],
    "nombre": "Carlos Perez",
    "presupuesto": "100",
    "tipo": "usuario",
    "usuario": "cPerez",
  },
]
`;
exports[`methods for managing clients and their needs must avoid removing non-existnig clients 1`] = `
Array [
  Object {
    "contraseña": "pass",
    "gasto": Array [],
    "nombre": "Ulises Gascón",
    "presupuesto": "1000",
    "tipo": "admin",
    "usuario": "ulises",
  },
  Object {
    "contraseña": "pass_2",
    "gasto": Array [],
    "nombre": "Carlos Perez",
    "presupuesto": "100",
    "tipo": "usuario",
    "usuario": "cPerez",
  },
]
`;
exports[`methods for managing clients and their needs must create new client 1`] = `
Array [
  Object {
    "contraseña": "pass",
    "gasto": Array [],
    "nombre": "Ulises Gascón",
    "presupuesto": "1000",
    "tipo": "admin",
    "usuario": "ulises",
  },
  Object {
    "contraseña": "pass_2",
    "gasto": Array [],
    "nombre": "Carlos Perez",
    "presupuesto": "100",
    "tipo": "usuario",
    "usuario": "cPerez",
  },
  Object {
    "contraseña": "pass2",
    "gasto": Array [],
    "nombre": "Teba Gomez",
    "presupuesto": "100",
    "tipo": "usuario",
    "usuario": "teba",
  },
]
`;
exports[`methods for managing clients and their needs must remove specific client 1`] = `
Array [
  Object {
    "contraseña": "pass_2",
    "gasto": Array [],
    "nombre": "Carlos Perez",
    "presupuesto": "100",
    "tipo": "usuario",
    "usuario": "cPerez",
  },
]
`;